### PR TITLE
Rework user management

### PR DESCRIPTION
**Pull request by svartkanin**
# Re-worked user management

## Fixes
- https://github.com/archlinux/archinstall/issues/1112
- https://github.com/archlinux/archinstall/issues/1183
- https://github.com/archlinux/archinstall/issues/1182
- https://github.com/archlinux/archinstall/issues/1109

### New file format for user credentials 
The format for the credentials config file has changed to allow specifying multiple users. This change is backwards compatible, so if old configurations get passed they are parseable and will internally be transformed into the new format. When saving configurations they will be saved in the new format. 

Old format
```
{
	"!root-password": "<root password>",
	"!users": {
		"username": {"!password": "<user password>"}
	},
	"!superusers": {
		"admin": {"!password": "<admin password>"}
	}
}
```

New format
```
{
	"!root-password": "<root password>",
	"!users": [
		{
			"username": "<USERNAME>",
			"!password": "<PASSWORD>",
			"sudo": false
		},
		{
			"username": "<SUDO_USERNAME>",
			"!password": "<PASSWORD>",
			"sudo": true
		}
	]
}
```

### User object
`!superuser` does not exist anymore, instead only the internal `!users` is handled whereas the entries are now a `User` class object 

```
@dataclass
class User:
	username: str
	password: str
	sudo: bool
```

This greatly simplified the code in some parts and makes it easier to avoid bugs :)

### Update to menu 

The menu has now only 1 entry to define users in which the sudo privileges can be specified. 

image

@Torxed I'll let you know when it's ready, probably tomorrow :)

### CODE REVIEW

**Comment by dylanmtaylor** [comment]
Can you also update the docs and schema for this change?

**Comment by svartkanin** [comment]
Yep will do that too :)

**Comment by Torxed** [comment]
I don't know why i did'nt do this right away haha, way more clean! Thank you!

**Comment by dylanmtaylor** [comment]
May I suggest we don't merge this until 2.5.0 since it's a schema-breaking change?

**Comment by svartkanin** [comment]
It's not a breaking change since it's backwards compatible :) 

I think the current state is way more fragile since we only support a single user in a file. Which means - and has been pointed out in the issues - that users will lose their actual settings if they try to export them.

In addition I think the 1 different settings of "users" and "superusers" seemed to have stirred quite some confusion which this resolves as well

File: archinstall/lib/installer.py [hunk 1]
```diff
@@ -1054,17 +1055,15 @@ def enable_sudo(self, entity: str, group :bool = False) -> bool:
 
 		return True
 
-	def user_create(self, user :str, password :Optional[str] = None, groups :Optional[str] = None, sudo :bool = False) -> None:
-		if groups is None:
-			groups = []
-
+	def user_create(self, users: List[User]) -> None:
```
Comment (Torxed) [inline archinstall/lib/installer.py:1058]: The function should stay as it were, in the sense that it was a function to perform one action.
It made it's purpose a bit more clear. And I propose we create a secondary function that deals with the loops and stuff, which would be called `load_users_layout()` or something similar to `load_disk_layout()`.
Thoughts?

**Comment by Torxed** [comment]
Yea it appears backwards compatible, I get that there's still some work left.
But I'd be happy to run tests on this on the backwards compatibility side and if they're successful I'd be happy to merge this in time for the release :)

File: archinstall/lib/installer.py (hunk 1 above)
Comment (svartkanin) [reply archinstall/lib/installer.py:1056]: Yes that's a great point! I'll update that one :)

**Comment by dylanmtaylor** [comment]
> Yea it appears backwards compatible, I get that there's still some work left. But I'd be happy to run tests on this on the backwards compatibility side and if they're successful I'd be happy to merge this in time for the release :)

I like this plan. I was under the impression that this was a breaking change

**Comment by svartkanin** [comment]
@Torxed This should be ready for a review now :)

**Comment by Torxed** [comment]
> @Torxed This should be ready for a review now :)

Testing commences now! :D

**Comment by Torxed** [comment]
It loads the old format where I created a `!superuser` and a `!user`.
screenshot

It also installs with the old configuration:
screenshot

And no shocker i guess, but the permissions are set correctly to:
screenshot

**Comment by Torxed** [comment]
Great work, especially making it backwards compatible. Merged and done! :)

**Comment by svartkanin** [comment]
Fantastic, thanks for double checking everything :)

**Comment by svartkanin** [comment]
I think the related issues where not closed automatically

**Comment by Torxed** [comment]
You are correct heh, gotta love Github.

**Comment by svartkanin** [comment]
I'm not sure what triggers it, maybe only mentioning it in the title 🤷‍♂️

**Comment by Torxed** [comment]
It's if the word `fix` is in the same sentence as the issue link I think.
There are some other aliases for it, but it has to be in the same sentence i recon.
But I don't mind closing the manually too hehe, gives me a sense of purpose :angel:

**Comment by erikdubois** [comment]
Could you add a text when we can start testing it?

**Comment by svartkanin** [comment]
It is in master now so basically from there now :), @Torxed will this make it into the new iso?

**Comment by Torxed** [comment]
> It is in master now so basically from there now :), @Torxed will this make it into the new iso?

It will :) I'm planning on running several tests throughout the day.
If they all succeed, I'm aiming to release the next version later this evening or very early tomorrow :)

**Pull request merged**